### ISSUE TITLE
[Roadmap] LMFlow Roadmap

### ISSUE THREAD

**Issue by wheresmyhair** [OptimalScale/LMFlow#862]
This document includes the features in LMFlow's roadmap. We welcome any discuss or contribute to the specific features at related Issues/PRs. 🤗 

### Main Features
* Data
	* [ ] DPO dataset format
	* [ ] Conversation template in DPO
	* [ ] jinja template
* Model
	* Backend
		* [ ] Accelerate support
* Pipeline
	* Train/Finetune/Align
		* [ ] DPO (multi-gpu)
		* [ ] Iterative DPO
		* [ ] PPO
		* [ ] LISA (multi-gpu, qwen2, chatglm)
		* [ ] Batch size and learning rate recommendation ([arxiv](https://arxiv.org/pdf/2405.14578))
	* Inference
		* [ ] vllm inference

### Usability
* [ ] Inference method auto-downgrading (vllm>ds, etc.), and make `vllm` package optional
* [ ] Merging similar model methods into `hf_model_mixin`

### Bug fixes
* [ ]  `model.generate()` with dsz3 #861 
* [ ] `merge_lora` lora with abs path merging

### Documentation
* [ ] Approx GPU memory requirement w.r.t model size & pipeline

**Comment by wheresmyhair** [comment]
Note on multiple instances vllm inference:
In vllm inference, the number of attn heads should be devisible by vllm tensor parallel size. If we have a 14 heads LLM, then the options for tp is 1&2 (7 will cause another division issue, but forget what that issue is). 
Say we have 8 gpus, then to utilize these devices, multiple instances vllm inference is necessary (tp=1 -> 8 instances, and tp=2 -> 4 instances)

**Comment by wheresmyhair** [comment]
Now supports Iterative DPO #883
